### ISSUE TITLE
How to get pending steps to not be seen as failures?

### ISSUE THREAD

**Issue by SteveDonie** [jenkinsci/cucumber-reports-plugin#129]
I've recently noticed that when a scenario has a step that returns 'pending' that the cucumber reports show the scenario as 'failed'. Is there a way to change that behavior? We use 'pending' to indicate that a certain dependency is not configured during a particular test run, and this is an expected condition. 

I have tried setting the report plugin configuration "Number of pending steps" to a very high number (99999) and that does not seem to affect the result.

**Comment by damianszczepanik** [comment]
This is done by design. What exactly are your expectations?

**Comment by SteveDonie** [comment]
I would expect the scenario to report scenarios that have pending steps to report their overall status as pending. 

The command-line reporting reports scenarios as failed, passed, or pending:
```
Failing Scenarios:
cucumber -p linux features/auditdb_and_reporting_mssql.feature:110 # Scenario: Verify a hammer insertSavedAuditData operation repeated, discrete H2 updates PROJECTS with exactly one row in MSSQL AuditDB (DVONE-4163)

3 scenarios (1 failed, 1 undefined, 1 passed)
171 steps (1 failed, 92 skipped, 1 undefined, 77 passed)
1m3.148s
```

The built-in html formatter also shows scenarios as passed, failed, or pending. I have attached a sample. 
[results.html.zip](https://github.com/jenkinsci/cucumber-reports-plugin/files/831512/results.html.zip)

I've also attached a screenshot showing what I am seeing with the cucumber-reports-plugin.
![screenshot from 2017-03-09 09-53-59](https://cloud.githubusercontent.com/assets/605418/23758202/a4dcc2de-04ae-11e7-9a81-03684efe7c85.png)

**Comment by SteveDonie** [comment]
Here's another html report with a single scenario that is pending.
[results.html.zip](https://github.com/jenkinsci/cucumber-reports-plugin/files/831553/results.html.zip)

**Comment by damianszczepanik** [comment]
Sorry but I don't see string enough reason for changing default behave

**Comment by SteveDonie** [comment]
Thanks. I'll probably just change it in my fork then. Would you mind pointing me towards where the logic to calculate the overall results are computed?

**Comment by damianszczepanik** [comment]
This should be good start point
https://github.com/damianszczepanik/cucumber-reporting/blob/master/src/main/java/net/masterthought/cucumber/json/Element.java#L115

**Comment by svsforsuccess** [comment]
But am planning to skip few scenarios so it would be greatly helpful if you add column for pending in scenarios and also if you could tell me the way to modify plugin code also should be fine...but i need it badly...thanks

**Comment by SteveDonie** [comment]
I have created a branch in my fork of cucumber-reporting that does almost all that is needed. The main reports now show pending and undefined steps/scenarios/features. The only thing I have not implemented is the trend reporting. 

![screenshot from 2017-06-13 10-14-09](https://user-images.githubusercontent.com/605418/27089896-4f1d5cbe-5021-11e7-829e-38e3ee500135.png)

If you'd like to get the code, see pull request 636: https://github.com/damianszczepanik/cucumber-reporting/pull/636

**Comment by mpkorstanje** [comment]
> I've recently noticed that when a scenario has a step that returns 'pending' that the cucumber reports show the scenario as 'failed'. Is there a way to change that behavior? We use 'pending' to indicate that a certain dependency is not configured during a particular test run, and this is an expected condition.

When executing cucumber tests in the strict mode, pending steps will fail the test so rendering pending as failures not strictly speaking wrong.

Semantically it is wrong to use Pending to mean not executed due to failing preconditions. However once cucumber-jvm 2.0.0 is out you'll be able to use [Junit Assume](http://junit.org/junit4/javadoc/4.12/org/junit/Assume.html) to do just this. 

Can't provide an eta. on that but we do have snapshot releases.

```
        <dependency>
            <groupId>io.cucumber</groupId>
            <artifactId>cucumber-junit</artifactId>
            <version>${cucumber.version}</version>
            <scope>test</scope>
        </dependency>

        <repository>
            <id>sonatype-snapshots</id>
            <url>https://oss.sonatype.org/content/repositories/snapshots</url>
            <snapshots>
                <enabled>true</enabled>
            </snapshots>
        </repository>
```

**Comment by svsforsuccess** [comment]
Thank you for update!!!

**Comment by svsforsuccess** [comment]
I have done woraround for this from excel ,am not including that feature
itself for run.

**Comment by maultosar** [comment]
This implementation seems quite reasonable, as Pending status tells us something was not yet implemented or is ommited on purpose. If something is wrong with stepdef(therefore scenario), there are failed and undefined statuses. So, +1 vote to see this in official implementation of plugin. 

We have our own BDD framework and we didn't have Pending exception yet, so I'm implementing it right now and it kind of depends on this plugin what statuses will I pass to scenario level.
So, is there any chance this will be implemented or I should not hope for this useful feature?

**Comment by svsforsuccess** [comment]
My requirement was skipping the feature if I do not want,But I implemented
it using pre requisite approach.

If any one interested in approach ,pls ping me here.

Thanks,
Karthik

**Comment by damianszczepanik** [comment]
Looks like there is need to add feature that was removed time ago because some problems with consistency. I need to think it over again

**Comment by maultosar** [comment]
It would be a very helpful feature. Thanks:)

**Comment by LucasSq** [comment]
This would be helpful for my project too :)

**Comment by jess8147** [comment]
+1 for this feature requirement

**Comment by KLBonn** [comment]
+1 from me

The result view shown by @SteveDonie ( [comment above](https://github.com/jenkinsci/cucumber-reports-plugin/issues/129#issuecomment-308154437) / [PR 636](https://github.com/damianszczepanik/cucumber-reporting/pull/636) ) would be exactly what we need.

Especially "undefined" should be differentiated from "failed": 
- "undefined" means for us: planned "work in progress", no special action implied.
- "failed" should be reserved for: Problems with application under test or test environment --> immediate action required!

**Comment by damianszczepanik** [comment]
Would it be OK to add tow new flags: one for undefined and second for pending so it defines if the steps should be counted as passed or failed?

**Comment by SteveDonie** [comment]
@damianszczepanik I don't think that your suggestion of new flags meets my needs at all. I don't want a feature or a scenario to be limited to a passed or failed state - I  actually like knowing that a scenario is 'pending' or 'undefined' - I see these as communicating more richly what the state of things is.

**Comment by damianszczepanik** [comment]
What when scenario has steps with different statuses?

**Comment by KLBonn** [comment]
Of course we have to have clearly defined aggregation rules.

We should bring the statuses into an order, and then take the worst (min or max) value.
E.g.:
- 0: passed
- 1: pending
- 2: undefined
- 3: failed

The outcome of a scenario then would simply be maximum of step results.

And same should hold for features which would be evaluated based on their worst scenario.


Am 21. Oktober 2017 11:01:10 MESZ schrieb Damian Szczepanik <notifications@github.com>:

**Comment by SteveDonie** [comment]
@KLBonn If you look at the diff in [pull request 636](https://github.com/damianszczepanik/cucumber-reporting/pull/636), you'll see that is exactly what I have done. 

See https://github.com/damianszczepanik/cucumber-reporting/pull/636/files#diff-8350411954262e85f9d3bee0c77b2879

**Comment by svsforsuccess** [comment]
Hi All,

Am not sure this is right tag,but

I would like to contribute following feature to this plugin ,please suggest


When I execute examples in feature in parallel

I have writtent an utility to merge all jsons according to this report and
show the result appropriately.

Thanks
Karthik

**Comment by damianszczepanik** [comment]
Based on https://github.com/damianszczepanik/cucumber-reporting/blob/master/src/main/java/net/masterthought/cucumber/json/support/Status.java#L17-L19 I'm not really sure if this is clear to everybody what is the relation between those three. So I have concern if by introducing three new statuses for scenarios and features we introduce some doubts what is the different between pending and skipped for feature. It is not as clear as for steps which is very small unit.

**Comment by KLBonn** [comment]
Maybe we could introduce a flag / feature toggle for that new behaviour?
So people could choose between "simple status" and "detailed status". 

Then anyone confused by differentiated statuses can drop down to showing every non-passing scenario/feature as failed.

Of course, a description / definition of the meaning for each status should be part of the documentation, anyway.

**Comment by kalyankp** [comment]
Hi ,
@damianszczepanik - I am still confused, with the present cucumber-report plugin how does ""Number of pending steps"  works? I have also tried giving 9999 in the "Number of pending steps" field , where as I have 7 pending steps in my Json. But still it overall status is failed.
Please suggest some quick solution for it.

Thank you.

**Comment by damianszczepanik** [comment]
Probably other statuses are set to zero while your report has more. Validate other statuses.

**Comment by kalyankp** [comment]
Hi, 
@damianszczepanik - Here I am attaching cucumber-report screenshot and configuration screenshot, Please see and let me know what mistake I am doing in configuration.

![cucumber-config](https://user-images.githubusercontent.com/33514225/33015828-2f9e1af8-ce12-11e7-96d7-1f1ff0e7793f.jpg)
![cucumber-report](https://user-images.githubusercontent.com/33514225/33015829-2fd545c8-ce12-11e7-9fd0-c7d290c5454a.jpg)

**Comment by damianszczepanik** [comment]
OK, now I understand, this is duplicate of https://github.com/damianszczepanik/cucumber-reporting/pull/636

**Comment by UltimateGeek** [comment]
The referenced pull request, damianszczepanik/cucumber-reporting#636 is going into the Datical fork, not jenkinsci. Why was this issue's status changed to closed?
